### PR TITLE
[CP] Add Ulysses context parallel

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4459
* #4608
* __->__ #4450

Add Ulysses CP with S(0)-to-S(1) all-to-all reshards around FlexAttention.

Keep its attention mask global, reject load balancing, and validate that attention heads are divisible by the combined TP and CP degree. Consolidate CP config validation and reject GPT-OSS because its per-head sinks are not sharded across CP.

Add unit and integration tests.

### CODE REVIEW

File: torchtitan/models/gpt_oss/parallelize.py [hunk 1]
```diff
@@ -65,6 +66,16 @@ def parallelize_gptoss(
...
 ):
+    # TODO(fegin): shard the per-head sinks over CP to support Ulysses.
+    if parallel_dims.cp_enabled and isinstance(
+        model.config.first_full_attention_backend, UlyssesCPFlexAttention.Config
+    ):
+        raise NotImplementedError(
+            "GPT-OSS does not support Ulysses context parallel because its "
+            "per-head sinks are sharded over TP but not CP. Use "
+            "AllGatherCPFlexAttention instead."
+        )
```
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/parallelize.py:77]: this validation on configs could have happened at config time, but yeah, not sure what's the best place to call it

File: torchtitan/models/common/cp_attention.py [hunk 2]
```diff
@@ -81,3 +83,41 @@ def forward(
...
         return super().forward(q_THK, k_THK, v_THV, **kwargs)
+
+
+class UlyssesCPFlexAttention(ContextParallelKernel, FlexAttention):
+    """Run FlexAttention with sequence-to-head all-to-all redistribution."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(FlexAttention.Config):
+        shard_attention_mask: ClassVar[bool] = False
+        shard_attention_heads: ClassVar[bool] = True
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py:94]: not used?

File: torchtitan/distributed/context_parallel/api.py [hunk 3]
```diff
@@ -114,6 +118,8 @@ def cp_shard(
     load_balancer_type: str | None = "headtail",
     input_seq_dims: int | tuple[int, ...] = 0,
     ptrr_mask_key: str | None = None,
+    *,
+    shard_attention_mask: bool = True,
```
Comment (tianyu-l) [inline torchtitan/distributed/context_parallel/api.py:122]: I feel this is too much indirection and hard to understand / maintain / extend.

A slightly more OOD way is to let each CPAttention backend own its own `cp_shard` method, and the Ulysses one would be different from the all-gather KV one.

Also there's implicit assumption that all layers have to use the same CP backend.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/cp_attention.py [hunk 4]
```diff
@@ -29,9 +30,11 @@
 __all__ = [
     "CPInnerAttention",
     "KVAllGatherCPFlexInnerAttention",
+    "UlyssesCPFlexInnerAttention",
 ]
 
 _SEQ_DIM = 0
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py]: maybe this should be renamed to `_TOKEN_DIM`, also aligning with the THK / THV notation

File: torchtitan/models/common/cp_attention.py [hunk 5]
```diff
@@ -102,3 +105,65 @@ def forward(
...
+            cp_mesh,
+            None,
+            None,
+        )
+        if attention_masks is not None:
+            input_dict["attention_masks"] = attention_masks
+        return input_dict
+
+    @staticmethod
+    def _reshard(
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py]: why do we need to introduce this function?

**Review by tianyu-l — APPROVED** [review]
please address inline comments
